nengo/builder/transforms.py

```diff
@@ -68,18 +65,9 @@ def build_sparse(model, transform, sig_in,
     assert weights.ndim == 2
 
     # Add operator for applying weights
-    if scipy_sparse:
-        assert isinstance(weights, scipy_sparse.spmatrix)
-        OperatorType = SparseDotInc
-    else:
-        assert isinstance(weights, np.ndarray)
-        OperatorType = DotInc
-        warnings.warn("Sparse Transform requires Scipy for sparse "
-                      "implementation. Falling back on dense matrix.")
-
     weight_sig = Signal(weights, name="%s.weights" % transform,
-                        readonly=True)
-    model.add_op(OperatorType(weight_sig, sig_in, weighted,
+                        readonly=True, type="sparse")
+    model.add_op(SparseDotInc(weight_sig, sig_in, weighted,
```

Comment:
So my worry with this is that weights may not actually be sparse (i.e. if Scipy is not installed). Then, this is confusing to backends, because we've got a Signal that says it's sparse and an operator that says it's a SparseDotInc, when neither of them are actually sparse.

This commit is supposed to make things easier for backends, but at least for NengoOCL (what I'm familiar with) it will make it harder. Before, if you didn't have Scipy installed, you'd get a dense Signal and a regular DotInc, which nengo-ocl supports and so SparseTransform would work without Scipy. However, with this commit, things will be expressed in sparse terms that nengo-ocl is not familiar with, and it will break.

The old code here was only making a sparse Signal and SparseDotInc if Scipy is installed, so I think it should have worked fine on backends.

Reply:
> So my worry with this is that weights may not actually be sparse (i.e. if Scipy is not installed). Then, this is confusing to backends, because we've got a Signal that says it's sparse and an operator that says it's a SparseDotInc, when neither of them are actually sparse.

I don't think it'd be confusing to backends, since the Signal explicitly tells the backend that they should treat this as a sparse signal (and same with the op).  The only tricky part is that in the case that the `initial_value` is not represented by a `scipy.sparse` format (e.g., because `scipy` is not installed), then the backend will have to manually convert the `initial_value` into the backend-specific sparse format.  But that seems like something that backends will probably have to do anyway.  So the only downside I can think of is that it might be slightly more difficult to convert a `numpy` array into the backend's sparse format than to convert a `scipy.sparse` array into that format, but I can't imagine there would be too much of a difference.

> This commit is supposed to make things easier for backends, but at least for NengoOCL (what I'm familiar with) it will make it harder. Before, if you didn't have Scipy installed, you'd get a dense Signal and a regular DotInc, which nengo-ocl supports and so SparseTransform would work without Scipy. However, with this commit, things will be expressed in sparse terms that nengo-ocl is not familiar with, and it will break.

If I'm understanding things correctly, the old behaviour only worked smoothly with NengoOCL if we assume that the user doesn't have scipy installed.  If the user does have scipy installed, then we would have the same problem ("things will be expressed in sparse terms that nengo-ocl is not familiar with, and it will break").  That doesn't seem like behaviour we'd want to rely on, so it seems like in either case NengoOCL will have to add logic to explicitly handle SparseDotInc (even if that behaviour is just to convert it to the equivalent dense DotInc).

With that in mind, this seems like an improvement for NengoOCL.  If you have support in NengoOCL for sparse signals and SparseDotInc, I presume that would be independent of `scipy`.  So in the old behaviour, if the user didn't have `scipy` installed, then their `SparseTransforms` would be converted to `DenseTransforms`, even though they could be running using NengoOCL's sparse support.  In the new behaviour, if the user doesn't have scipy installed then the same data is available to NengoOCL (the dense equivalent of the sparse transform).  So if you wanted, you could just implement that with a regular DotInc, same as before.  But you could also notice that this is a sparse op (even though the initial value isn't sparse), and take advantage of that to use the sparse support in NengoOCL.

> The old code here was only making a sparse Signal and SparseDotInc if Scipy is installed, so I think it should have worked fine on backends.

It works, but it requires scipy to be installed to work.  Unless backends are actually using `scipy` to do the implementation, it seems awkward to force `scipy` to be installed just to activate that functionality.

Reply:
Those are some fair points. I had been treating the fall-back to Numpy as just to allow simple examples to not break for users, not as something with more practical applications.

In the long run, I do worry about representing sparse signals with dense matrices, for more than just size concerns. One issue comes up when you get to learning, where you may have a sparse matrix where you want to adjust the values with a learning rule, while maintaining the sparsity structure. In that case, the Numpy matrix won't necessarily define the sparsity structure, if e.g. you start with zeros for some elements but allow those elements to change.

In the short term, my biggest problem with this is just that it makes it less predictable what the data format in sparse Signals will be. It's obviously not hard for backends to check and deal with this accordingly, it's just another thing they have to be aware of.

Also, I just realised it makes the API a little weird. As things are, a user could not pass in a Numpy array with zeros and have it be treated sparsely by a backend (`Sparse.init` has to be the same length as `Sparse.inds` if it's a Numpy array). We could change this so that if you pass `init` without `inds`, it can be a Scipy sparse matrix or a Numpy matrix.

Personally, I don't have a problem with requiring Scipy to do anything practical with Sparse transforms, and I would slightly lean to that side just to simplify things a bit for backends (they don't have to interpret a Numpy array as a sparse matrix). But I'm also fine with these changes if that's the direction that others are leaning. Maybe we should do a straw poll.

Reply:
Are there examples anywhere of models that would make use of sparse transforms? I feel like right now there are lots of hypothetical situations being discussed and if those were made into full examples (which we probably want for our documentation anyway) it would make the decision pretty easy.

Reply:
> Also, I just realised it makes the API a little weird. As things are, a user could not pass in a Numpy array with zeros and have it be treated sparsely by a backend (Sparse.init has to be the same length as Sparse.inds if it's a Numpy array). We could change this so that if you pass init without inds, it can be a Scipy sparse matrix or a Numpy matrix.

I think we want users to always be specifying things in a sparse format if they want to use a SparseTransform (either by using `scipy.sparse`, or by specifying data/`inds`).  

> One issue comes up when you get to learning, where you may have a sparse matrix where you want to adjust the values with a learning rule, while maintaining the sparsity structure. In that case, the Numpy matrix won't necessarily define the sparsity structure, if e.g. you start with zeros for some elements but allow those elements to change.

Yeah that is true, the numpy fallback assumes that the sparsity structure won't change (so you can detect it just by searching for non-zero elements). But we don't support learning on `SparseTransforms` at the moment (and I'm not sure we would any time soon), so I'd say that's something we can worry about if/when the time comes.

> Are there examples anywhere of models that would make use of sparse transforms? I feel like right now there are lots of hypothetical situations being discussed and if those were made into full examples (which we probably want for our documentation anyway) it would make the decision pretty easy.

Really none of this makes a difference from the user's perspective, so I don't think the difference would show up in any examples.

The difference we're focusing on in this commit is how data is represented in-between the time the user specifies a ` SparseTransform` and when it is delivered to the backend.  Basically, whether we want to have a `scipy` bottleneck in-between the user and the backend (where the data is represented internally as `scipy` data structures during the build process).  The advantage of that is that we get a sparse matrix representation for free, which naturally stores the data we need to represent.  The disadvantage is that the user needs to have `scipy` installed, even if they're not using `scipy` to specify the `SparseTransform` and the backend isn't using `scipy` to implement it.

Reply:
https://github.com/nengo/nengo-extras/blob/master/docs/examples/mnist_single_layer.ipynb

I'm not sure if it clarifies things, though. A lot of what we're talking about here is on the backend side, not the frontend API.

Reply:
> The disadvantage is that the user needs to have scipy installed, even if they're not using scipy to specify the SparseTransform and the backend isn't using scipy to implement it.

I had thought originally about essentially defining our own sparse format for the sparse Signals, like the data/inds format we accept for creating a sparse transform. Then I realised that this would be re-inventing the wheel, since `scipy.sparse` already does this, but in a more flexible and efficient manner. So I opted to go with Scipy's sparse matrix as our internal representation. But you're right, that it does mean we're tied to it even if neither the user nor the backend are using it.

Reply:
Just had a brief chat with Dan and Trevor. Dan is going to try implementing the data/inds format in `Signal`. However, if the user passes in a Scipy sparse matrix, we'd preserve that. So the data in a sparse `Signal` could either be a Scipy sparse matrix or data/inds. Then the operator (i.e. `SparseDotInc`) implementation can do what it wants, like pumping data/inds straight to whatever sparse format the backend uses for simulation. Thus, we don't require Scipy to do sparse stuff.